src/components/AccountDrawer/MiniPortfolio/PortfolioLogo.tsx

```diff
@@ -87,81 +78,114 @@ const L2LogoContainer = styled.div<{ $backgroundColor?: string }>`
+  logo2?: string
+  size: string
+  onError1?: () => void
+  onError2?: () => void
+}
+
+function DoubleLogo({ logo1, onError1, logo2, onError2, size }: DoubleLogoProps) {
+  return (
+    <DoubleLogoContainer>
+      <CircleLogoImage size={size} src={logo1 ?? blankTokenUrl} onError={onError1} />
+      <CircleLogoImage size={size} src={logo2 ?? blankTokenUrl} onError={onError2} />
+    </DoubleLogoContainer>
+  )
+}
+
+type DoubleCurrencyLogoProps = {
+  chainId: ChainId
+  currencies: Array<Currency | undefined>
+  backupImages?: (string | undefined)[]
+  size: string
+}
+
+function DoubleCurrencyLogo({ chainId, currencies, backupImages, size }: DoubleCurrencyLogoProps) {
+  const [src, nextSrc] = useTokenLogoSource(
+    currencies?.[0]?.wrapped.address,
+    chainId,
+    currencies?.[0]?.isNative,
+    backupImages?.[0]
+  )
+  const [src2, nextSrc2] = useTokenLogoSource(
+    currencies?.[1]?.wrapped.address,
+    chainId,
+    currencies?.[1]?.isNative,
+    backupImages?.[1]
+  )
+
+  return currencies.length > 1 ? (
+    <DoubleLogo logo1={src} onError1={nextSrc} logo2={src2} onError2={nextSrc2} size={size} />
+  ) : currencies.length === 1 && src ? (
+    <CircleLogoImage size={size} src={src} onError={nextSrc} />
+  ) : (
+    <MissingImageLogo size={size}>
+      {currencies[0]?.symbol?.toUpperCase().replace('$', '').replace(/\s+/g, '').slice(0, 3)}
+    </MissingImageLogo>
+  )
+}
+
+function PortfolioAvatar({ accountAddress, size }: { accountAddress: string; size: string }) {
   const { avatar, loading } = useENSAvatar(accountAddress, false)
-  const theme = useTheme()
 
-  const [src, nextSrc] = useTokenLogoSource(currencies?.[0]?.wrapped.address, chainId, currencies?.[0]?.isNative)
-  const [src2, nextSrc2] = useTokenLogoSource(currencies?.[1]?.wrapped.address, chainId, currencies?.[1]?.isNative)
+  return loading ? (
+    <Loader size={size} />
+  ) : avatar ? (
+    <ENSAvatarImg src={avatar} alt="avatar" />
+  ) : (
+    <Unicon size={40} address={accountAddress} />
+  )
+}
+
+type PortfolioLogoProps = {
+  chainId: ChainId
+  accountAddress?: string
+  currencies?: Array<Currency | undefined>
+  images?: (string | undefined)[]
+  size?: string
+  style?: React.CSSProperties
+}
+
+function SquareL2Logo({ chainId }: { chainId: ChainId }) {
+  if (chainId === ChainId.MAINNET) return null
+  const { squareLogoUrl, logoUrl } = getChainInfo(chainId)
 
-  let component
+  const chainLogo = squareLogoUrl ?? logoUrl
+
+  return (
+    <L2LogoContainer hasChainLogo={!!squareLogoUrl}>
+      {squareLogoUrl ? (
+        <SquareChainLogo src={chainLogo} alt="chainLogo" />
+      ) : (
+        <StyledChainLogo src={chainLogo} alt="chainLogo" />
+      )}
+    </L2LogoContainer>
+  )
+}
+
+/**
+ * Renders an image by prioritizing a list of sources, and then eventually a fallback contract icon
+ */
+export function PortfolioLogo({ chainId, accountAddress, currencies, images, size = '40px' }: PortfolioLogoProps) {
+  let logo
```

Comment:
nit: you can make this a `const` if you extract the conditionals to a function and just to a bunch of checks and returns - might be nicer